llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

```diff
@@ -418,10 +513,36 @@ MachineInstr *SDWASrcOperand::potentialToConvert(const SIInstrInfo *TII,
     return nullptr;
 
   MachineInstr *Parent = PotentialMO->getParent();
+  if (Parent->isRegSequence()) {
+    Parent = regSequenceFindSingleSubregUser(
+        *Parent, getReplacedOperand()->getReg(), getMRI());
+    return Parent && canCombineSelections(*Parent, TII) ? Parent : nullptr;
+  }
 
   return canCombineSelections(*Parent, TII) ? Parent : nullptr;
 }
 
+/// Returns true if \p RHS is either the same register as LHS or the
+/// defining instruction of \p LHS is a REG_SEQUENCE in which \p
+/// RHS occurs as the operand for the register that corresponds to the
+/// subregister of LHS.
+static bool isSameRegOrCopy(const MachineOperand &LHS,
+                            const MachineOperand &RHS,
+                            const MachineRegisterInfo *MRI) {
+  if (isSameReg(LHS, RHS))
+    return true;
+
+  const MachineOperand *Def = findSingleRegDef(&LHS, MRI);
+  const MachineInstr *MI = Def ? Def->getParent() : nullptr;
+
+  // TODO Handle other copy-like instructions?
+  if (!MI || !MI->isRegSequence())
+    return false;
+
+  auto SubReg = regSequenceFindSubreg(*MI, RHS.getReg());
+  return SubReg && LHS.getSubReg() == SubReg;
```

Comment:
If all you're doing is looking through copy-like instructions, perhaps this pass should not be directly responsible for this. I think either peephole-opt or SIFoldOperands should be pulling the subregister read through the reg_sequence. This change isn't core to the SDWA problem.


i.e.

```
    %4:vreg_64 = REG_SEQUENCE %1, %subreg.sub0, %3, %subreg.sub1
    %7:vreg_64 = REG_SEQUENCE %6, %subreg.sub0, %3, %subreg.sub1
    %8:vgpr_32, %9:sreg_64_xexec = V_ADD_CO_U32_e64 %4.sub0, %7.sub0, 0, implicit $exec
```

Should have been rewritten to directly read the reg_sequence inputs in the first place 

```
    %8:vgpr_32, %9:sreg_64_xexec = V_ADD_CO_U32_e64 %1, %6, 0, implicit $exec
```